### PR TITLE
Added button to payout bounty

### LINKED ISSUE #8: better interface for bounty-payout
# Bounty

the current way to talk the [bounty payout contract](https://github.com/leapdao/leap-contracts/blob/master/contracts/misc/BountyPayout.sol) is very unintuitive.

Screen Shot 2019-10-23 at 4 52 27 PM

gardeners often forget to add 18 zeros to the end of amounts, or don't know how to convert string to hex for the issueID (last field).

## Scope

- add additional button next to "Add DAI tx"
- create an input field that can work for `payout()`, `payoutNoWorker()`, and `payoutNoReviewer()` functions
- automatically add zeros and convert string to hex

## Deliverables

- code
- deployed at [wallet.leapdao.org](https://wallet.leapdao.org)

## Gain for the project

- better integration with colony network contracts

### Publicly visible delivery

- blog post about new UI + colony integration

## Roles
bounty gardener: @troggy  / 50 DAI flat
bounty worker: @eswarasai / 85%, 425 DAI
bounty reviewer: name / 15%, 75 DAI

### PR DESCRIPTION
**Pull request by eswarasai**
### Description
- [x] Added interface for the ease of paying out bounties
- [x] Handle different scenarios of paying out bounties i.e., `gardener`/`worker`/`reviewer`
- [x] Added reputation only checkbox
- [x] Create a multisig transaction with required input parameters to payout bounty

Fixes: #8

### CODE REVIEW

**Comment by eswarasai** [comment]
@troggy - I think this is ready for initial review. Few of the things are missing from below:

- Add validation for `worker` and `reviewer` address
- Not sure if we need to add `isRepOnly` checkbox and also how it fits in exactly

Let me know your initial review/thoughts, so that I can quickly iterate. Thanks!

**Comment by troggy** [comment]
thanks, I will review 👍

> Not sure if we need to add isRepOnly checkbox and also how it fits in exactly

Yes, it is needed. Sometimes we want just to register reputation, not actual payout. I can imagine that `Reputation only` checkbox should affect transaction like that:
- If "Reputation only" checkbox is unchecked, do nothing
- if "Reputation only" checkbox is checked then dapp should add 1 wei to the amount (no need to reflect amount change on the UI of course)

For instance:
200 DAI + "Reputation only" checked → 200 × 10<sup>18</sup> **+ 1** is sent in transaction
200 DAI + "Reputation only" unchecked → 200 × 10<sup>18</sup> is sent in transaction

**Comment by eswarasai** [comment]
@troggy - Thanks for the detailed explanation. I'll incorporate the above change as well 👍 

One more thing to note is, I'm only using `payout()` method of the `BountyPayout` contract as the UI handles the default `gardener`, `worker` and `reviewer` fields. Moreover if the amount of `worker`/`reviewer` is 0, the actual payout isn't sent anyway. I hope this behaviour is fine. Thanks!

**Comment by troggy** [comment]
@eswarasai I've fixed an error in my previous message:
In example with 200 reputation payout correct value is
```
200 × 1000000000000000000 + 1
```

**Comment by eswarasai** [comment]
@troggy - No worries. I got the gist of what should be done. I've been going through the `leap-contracts` PR for Bounty Payout V2 as well for the reference to ensure I'm implementing things right 😅

**Comment by eswarasai** [comment]
@troggy - Added the reputation only checkbox as well 🙂

File: dapp/src/controllers/payoutBountyCtrl.js [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+              stripHexPrefix(toHex(_isRepOnly ? _amountBN.add(1).toString() : _amountBN.toString())),
+              24
+            );
+            return `0x${stripHexPrefix(_address)}${_amountHex}`
+          };
+
+          const _gardener = _getInput($scope.gardener.address, $scope.gardener.amount, $scope.gardener.isRepOnly);
+          const _worker = _getInput($scope.worker.address, $scope.worker.amount, $scope.worker.isRepOnly);
+          const _reviewer = _getInput($scope.reviewer.address, $scope.reviewer.amount, $scope.reviewer.isRepOnly);
+          const _bountyId = toHex($scope.bountyId.replace('github.com/leapdao', ''));
```
Comment (troggy) [inline dapp/src/controllers/payoutBountyCtrl.js]:
```suggestion
          const _bountyId = toHex($scope.bountyId.replace('https://github.com/leapdao', ''));
```

**Comment by johba37** [comment]
# filled out this payment:

Screen Shot 2019-11-02 at 4 58 23 PM

## expected:

function `payoutReviewedDelivery()` would be called.

## actual:

function `payout()` was called:
Screen Shot 2019-11-02 at 4 58 44 PM

**Review by johba37 — CHANGES_REQUESTED** [review]
incorrect function called here?
https://github.com/leapdao/MultiSigWallet/pull/9#issuecomment-549056626

**Comment by troggy** [comment]
@johannbarbie that was me who confused @eswarasai to use `payout()` for any payment (it is simpler). I wasn't aware of gas savings you get if you use `payoutReviewedDelivery()` back then

**Comment by eswarasai** [comment]
@johannbarbie - For the ease of it, I'm just calling the `payout` method even when there's no worker/reviewer with their respective amounts `0` since the payment won't happen. In case you'd like to have better transaction history, I can do the necessary handling of calling the particular method in case of each scenario.

Just to be sure I don't misunderstand the use cases, please do list which function to be called in each scenario. Thanks 🙂

**Comment by johba37** [comment]
> @johannbarbie that was me who confused @eswarasai to use `payout()` for any payment (it is simpler). I wasn't aware of gas savings you get if you use `payoutReviewedDelivery()` back then

because of the colony overhead, it is quite a lot, 500,000 🌻 

`payout()` [1.5M](https://etherscan.io/tx/0xa48d71d36ae6489fcbbae38538fbb256a86cbc8b2c7afba9c406402c5df112c3)
`payoutNoReviewer()` [1.0M](https://etherscan.io/tx/0x3642f2b227ba75a84f93055ba896ac55089bdbf2051dc20d0d567165f6c81e16)

at 7.5gWei that is **0.6 USD saving in 1 TX !!!**

**Comment by johba37** [comment]
> please do list which function to be called in each scenario

gardener + worker + reviewer => `payout()`
gardener + reviewer =>  `payoutReviewedDelivery()`
gardener + worker => `payoutNoReviewer()`

**Comment by eswarasai** [comment]
@johannbarbie - I see that you've already tested the code on `mainnet`? Apart from the above method change, is there anything else needed to be handled?

I'm wondering if we need the below or not? Let me know if it's of priority.
- Add validation for worker and reviewer address

cc: @troggy

**Comment by troggy** [comment]
> Add validation for worker and reviewer address

it should be easy to add. Otherwise, not a priority — the tool is for internal use. IMO

**Comment by johba37** [comment]
> I see that you've already tested the code on mainnet? 

yeah, here: https://etherscan.io/tx/0x228b47e236caada562f9d7a31e57ccc98f972d360c741c2ae4716e1febafb100

> Apart from the above method change, is there anything else needed to be handled?

in the `bountyURL` field, there is no length check.
if people enter `https://github.com/leapdao/MultiSigWallet/issues/8`
then only `/MultiSigWallet/issues/8` should be converted to hex (everything after leapdao):
=> `0x2f4d756c746953696757616c6c65742f6973737565732f38`
you are doing that already correctly. 👍 

the bountyURL field is only 32 bytes in length, so if the part after `https://github.com/leapdao` is longer than 32 characters, user should see an error. currently i am able to submit longer values, and the tx will fail.

**Comment by eswarasai** [comment]
@johannbarbie - Thanks for the above comment. I'll add the 32 characters length check as well.

> payout() 1.5M
> payoutNoReviewer() 1.0M

Looking at these values, I'm afraid I need to increase the `gas` for the transaction options? Right now it's just `500,000`. Let me know what should be the `gas` limits for all the 3 methods. Thanks! 🙂

**Comment by johba37** [comment]
> Looking at these values, I'm afraid I need to increase the `gas` for the transaction options? Right now it's just `500,000`. Let me know what should be the `gas` limits for all the 3 methods. Thanks! 🙂

thx, don't worry about that 🙂 . the values i've listed are for executing transactions. initializing transactions is way cheaper.

**Comment by eswarasai** [comment]
@johannbarbie - So based on the below scenarios, I'm assuming that there'll always be either `worker` or `reviewer` along with `gardener` for bounty payout?

> gardener + worker + reviewer => payout()
> gardener + reviewer => payoutReviewedDelivery()
> gardener + worker => payoutNoReviewer()

In case both `worker` & `reviewer` details are not present, would it be alright if I don't allow the Payout to happen? Thanks!

**Comment by eswarasai** [comment]
@johannbarbie - Updated the PR with all the validations as well as calling respective methods in case of no worker/reviewer. Please do test and let me know if there's anything else pending.

Before merging, if you can let me know the deployed Mainnet address of the Bounty Payout V2 contract, I'll do a quick commit for that as well. Thanks!

cc: @troggy

**Comment by johba37** [comment]
> In case both `worker` & `reviewer` details are not present, would it be alright if I don't allow the Payout to happen? Thanks!

in LeapDAO any payouts is supposed to pass by 2 pairs of eyes, hence, in addition to gardener, either worker or reviewer has to be present :)

**Comment by johba37** [comment]
for max value i saw you used `999999999999999`, our [biggest bounty](https://app.glassfrog.com/organizations/14849/orgnav/policies/10912780) is 1400 DAI, so we could use that.

**Review by johba37 — APPROVED** [review]
great work

**Comment by eswarasai** [comment]
@johannbarbie - Thanks for the approval. I can update the max value for amount field to 1400 but I also see the below statement in the policy:

> If 2 people work on the bounty together, the payout increases by 1.5x.

Should I increase the max value to `1.5x 1400 = 2100` then?

**Comment by johba37** [comment]
> Should I increase the max value to `1.5x 1400 = 2100` then?

yes, great point 👍

**Comment by eswarasai** [comment]
@johannbarbie - Updated. Let me know if there's anything else needed or good to be merged 🙂

**Pull request merged**